packages/lit-html/src/directives/repeat.ts

```diff
@@ -0,0 +1,458 @@
+        );
+        oldHead++;
+        newTail--;
+      } else if (oldKeys[oldTail] === newKeys[newHead]) {
+        // Old tail matches new head; update and move to new head
+        newParts[newHead] = setPartValue(
+          oldParts[oldTail]!,
+          newValues[newHead]
+        );
+        insertPartBefore(containerPart, oldParts[oldTail]!, oldParts[oldHead]!);
+        oldTail--;
+        newHead++;
+      } else {
+        if (newKeyToIndexMap === undefined) {
+          // Lazily generate key-to-index maps, used for removals &
+          // moves below
+          newKeyToIndexMap = generateMap(newKeys, newHead, newTail);
+          oldKeyToIndexMap = generateMap(oldKeys, oldHead, oldTail);
+        }
+        if (!newKeyToIndexMap.has(oldKeys[oldHead])) {
+          // Old head is no longer in new list; remove
+          removePart(oldParts[oldHead]!);
+          oldHead++;
+        } else if (!newKeyToIndexMap.has(oldKeys[oldTail])) {
+          // Old tail is no longer in new list; remove
+          removePart(oldParts[oldTail]!);
+          oldTail--;
+        } else {
+          // Any mismatches at this point are due to additions or
+          // moves; see if we have an old part we can reuse and move
+          // into place
+          const oldIndex = oldKeyToIndexMap.get(newKeys[newHead]);
+          const oldPart = oldIndex !== undefined ? oldParts[oldIndex] : null;
+          if (oldPart === null) {
+            // No old part for this value; create a new one and
+            // insert it
+            const newPart = createAndInsertPart(
+              containerPart,
+              oldParts[oldHead]!
+            );
+            setPartValue(newPart, newValues[newHead]);
+            newParts[newHead] = newPart;
+          } else {
+            // Reuse old part
+            newParts[newHead] = setPartValue(oldPart, newValues[newHead]);
+            insertPartBefore(containerPart, oldPart, oldParts[oldHead]!);
+            // This marks the old part as having been used, so that
+            // it will be skipped in the first two checks above
+            oldParts[oldIndex as number] = null;
+          }
+          newHead++;
+        }
+      }
+    }
+    // Add parts for any remaining new values
+    while (newHead <= newTail) {
+      // For all remaining additions, we insert before last new
+      // tail, since old pointers are no longer valid
+      const newPart = createAndInsertPart(containerPart, newParts[newTail + 1]);
+      setPartValue(newPart, newValues[newHead]);
+      newParts[newHead++] = newPart;
+    }
+    // Remove any remaining unused old parts
+    while (oldHead <= oldTail) {
+      const oldPart = oldParts[oldHead++];
+      if (oldPart !== null) {
+        removePart(oldPart);
+      }
+    }
+
+    // Save order of new parts for next round
+    this.itemKeys = newKeys;
+    // Directly set part value, bypassing it's dirty-checking
+    containerPart._value = newParts;
```

Comment:
Add a TODO linked to #1261

Reply:
Also note that this doesn't strictly need to use `_value` to store the `newParts`. The only actual requirement is that first render try `getPartValue` to get hydrated parts; after that, it can prefer an instance variable on the part to save/retrieve the previously rendered parts.

Reply:
That's true, but might as well reuse that storage, and this makes it work that way for hydration.